File: scripts/cartographer/hforge-agent-hook.py
#!/usr/bin/env python3
"""Recommend-only Harness Forge agent hook helper.

Prints the command plan for an agent lifecycle event. It never executes
commands; execution and audit happen in-process via ``hforge agent hook``.
"""
from __future__ import annotations

import argparse
import json
import os
from dataclasses import dataclass, asdict
from typing import List


@dataclass(frozen=True)
class Recommendation:
    command: str
    reason: str
    autoExecutable: bool = True


def q(value: str) -> str:
    escaped = value.replace('"', '\\"')
    return f'"{escaped}"'


def split_files(value: str | None) -> List[str]:
    if not value:
        return []
    return [item.strip() for item in value.split(',') if item.strip()]


def build_plan(event: str, goal: str, files: List[str], hforge: str) -> List[Recommendation]:
    file_arg = ','.join(files)
    recs: List[Recommendation] = []
    if event == 'task.started':
        recs.append(Recommendation(f'{hforge} graph build --if-stale', 'Keep project graph fresh.'))
        recs.append(Recommendation(f'{hforge} context compile --goal {q(goal or "")}', 'Produce focused task context.'))
    elif event == 'task.context_needed':
        recs.append(Recommendation(f'{hforge} context compile --goal {q(goal or "")}', 'Agent requested focused context.'))
    elif event in ('files.pre_edit', 'files.changed'):
        recs.append(Recommendation(f'{hforge} impact --files {q(file_arg)} --json', 'Estimate change impact.'))
    elif event in ('command.failed', 'tests.failed'):
        recs.append(Recommendation(f'{hforge} impact --files {q(file_arg)} --json', 'Link failure to impacted modules.'))
    elif event in ('pr.prep', 'task.completed'):
        recs.append(Recommendation(f'{hforge} impact --changed --json', 'Summarize changed-file impact.'))
    else:
        recs.append(Recommendation(f'{hforge} context compile --goal {q(goal or event)}', 'Fallback context.', False))
    return recs


def main() -> int:
    parser = argparse.ArgumentParser(description='Recommend-only Harness Forge agent hook helper')
    parser.add_argument('--event', required=True)
    parser.add_argument('--goal', default='')
    parser.add_argument('--files', default='')
    parser.add_argument('--command', default='')
    parser.add_argument('--log', default='')
    parser.add_argument('--execute', action='store_true')
    parser.add_argument('--json', action='store_true')
    args = parser.parse_args()

    hforge = os.environ.get('HFORGE_BIN', 'hforge')
    files = split_files(args.files)
    recs = build_plan(args.event, args.goal, files, hforge)
    diagnostics: List[str] = []
    if args.execute:
        diagnostics.append(
            "this helper never executes commands; run 'hforge agent hook --execute' for in-process diagnostic execution"
        )

    payload = {
        'event': args.event,
        'mode': 'dry-run',
        'recommendedCommands': [asdict(r) for r in recs],
        'executedCommands': [],
        'diagnostics': diagnostics,
        'nextAction': "review the plan, then call 'hforge agent hook' for execution and audit",
    }

    if args.json:
        print(json.dumps(payload, indent=2))
    else:
        print(f'Event: {args.event}')
        print('Mode: dry-run')
        for item in diagnostics:
            print(f'Note: {item}')
        print('Recommended commands:')
        for rec in recs:
            print(f'- {rec.command}')
            print(f'  reason: {rec.reason}')
    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: scripts/cartographer/hforge-agent-hook.mjs
#!/usr/bin/env node
function arg(name, fallback = "") {
  const idx = process.argv.indexOf(`--${name}`);
  if (idx >= 0 && idx + 1 < process.argv.length) {
    return process.argv[idx + 1];
  }
  return fallback;
}

function hasFlag(name) {
  return process.argv.includes(`--${name}`);
}

function q(value) {
  return `"${String(value || "").replace(/"/g, '\\"')}"`;
}

function rec(command, reason, autoExecutable = true) {
  return { command, reason, autoExecutable };
}

function buildPlan(event, goal, files, hforge) {
  const recs = [];
  if (event === "task.started") {
    recs.push(rec(`${hforge} graph build --if-stale`, "Keep project graph fresh."));
    recs.push(rec(`${hforge} context compile --goal ${q(goal)}`, "Produce focused task context."));
  } else if (event === "task.context_needed") {
    recs.push(rec(`${hforge} context compile --goal ${q(goal)}`, "Agent requested focused context."));
  } else if (event === "files.pre_edit" || event === "files.changed") {
    recs.push(rec(`${hforge} impact --files ${q(files)} --json`, "Estimate change impact."));
  } else if (event === "command.failed" || event === "tests.failed") {
    recs.push(rec(`${hforge} impact --files ${q(files)} --json`, "Link failure to impacted modules."));
  } else if (event === "pr.prep" || event === "task.completed") {
    recs.push(rec(`${hforge} impact --changed --json`, "Summarize changed-file impact."));
  } else {
    recs.push(rec(`${hforge} context compile --goal ${q(goal || event)}`, "Fallback context.", false));
  }
  return recs;
}

const event = arg("event", "task.started");
const goal = arg("goal", "");
const files = arg("files", "");
const asJson = hasFlag("json");
const hforge = process.env.HFORGE_BIN || "hforge";
const recommendedCommands = buildPlan(event, goal, files, hforge);
const diagnostics = hasFlag("execute")
  ? ["this helper never executes commands; run 'hforge agent hook --execute' for in-process diagnostic execution"]
  : [];
const payload = {
  event,
  mode: "dry-run",
  recommendedCommands,
  executedCommands: [],
  diagnostics,
  nextAction: "review the plan, then call 'hforge agent hook' for execution and audit",
};

if (asJson) {
  process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
} else {
  process.stdout.write(`Event: ${event}\nMode: dry-run\nRecommended commands:\n`);
  for (const item of recommendedCommands) {
    process.stdout.write(`- ${item.command}\n  reason: ${item.reason}\n`);
  }
}
